### PR TITLE
kubelet: report NodeReady last in status list

### LINKED ISSUE #16961: Unable to schedule pods in mixed version cluster (scheduler=1.0.7, kubelet=1.1.1-beta.1)
When testing a mixed-version cluster it seems that pods will not be scheduled due to an issue with how multiple node conditions are evaluated.

In this case kubelet=v1.1.1-beta.1 and scheduler (and all other components) are v1.0.7.

It seems the issue is surfaced by adding a second node condition (and was also resolved) here:  https://github.com/kubernetes/kubernetes/commit/a2570926da1b87856351c783c908c255042a9193#diff-53f28720440ab4b589ef2273346cecfcL139

The v1.0.7 code will always evaluate the status of the last condition in the list (now "OutOfDisk=False"), when it should be checking the "Ready" condition. So now, "Ready" always evaluates to "False".

Then pods fail to schedule with "no nodes available to schedule pods", although `kubectl get nodes` reports nodes as ready.

Not sure if this type of mixed-version use is within the bounds of reasonable, but I'll submit a simple (backported) fix from the above commit in case.

### PR DESCRIPTION
**Pull request by aaronlevy**
fixes: https://github.com/kubernetes/kubernetes/issues/16961

/cc @bgrant0607

### CODE REVIEW

**Comment by bgrant0607** [comment]
cc @quinton-hoole @ihmccreery

**Comment by bgrant0607** [comment]
Assigned to @dchen1107 to choose a reviewer.

**Comment by ncdc** [comment]
cc @kubernetes/rh-cluster-infra

**Comment by pmorie** [comment]
@k8s-bot ok to test

**Comment by pmorie** [comment]
Looks like a weird error bringing up the cluster.

@k8s-bot test this

**Comment by aaronlevy** [comment]
Hmm. I ran some manual tests locally, but looks like this mucked up the ready state. I'll try again on my end, maybe I messed up my tests.

**Comment by aaronlevy** [comment]
Yeah, my tests were against mixed release versions (worker kubelet with these changes and v1.0.7/v1.1.1 master components). I'll take a look tomorrow why this might be broken against HEAD.

**Comment by aaronlevy** [comment]
The test was expecting the "Ready" node condition to be the first in the list of conditions. Added a change to check specifically for the condition by name.

File: cluster/validate-cluster.sh [hunk 1]
```diff
@@ -40,7 +40,7 @@ while true; do
   # Suppress errors from kubectl output because during cluster bootstrapping
   # for clusters where the master node is registered, the apiserver will become
   # available and then get restarted as the kubelet configures the docker bridge.
-  nodes_status=$("${KUBE_ROOT}/cluster/kubectl.sh" get nodes -o template --template='{{range .items}}{{with index .status.conditions 0}}{{.type}}:{{.status}},{{end}}{{end}}' --api-version=v1) || true
+  nodes_status=$("${KUBE_ROOT}/cluster/kubectl.sh" get nodes -o template --template='{{range .items}}{{range .status.conditions}}{{if eq .type "Ready"}}{{.type}}:{{.status}},{{end}}{{end}}{{end}}' --api-version=v1) || true
```
Comment (pmorie) [inline cluster/validate-cluster.sh]: trivia: If you want to be make this more idiomatic bash (ie, less readable to normal human beings), the bashism for this is `|| :`

**Comment by pmorie** [comment]
LGTM, does this ring the bell for you too @dchen1107 ?

**Comment by pmorie** [comment]
@k8s-bot ok to test

**Comment by pmorie** [comment]
@k8s-bot test this

**Comment by dchen1107** [comment]
LGTM

Do we expect to carry this for a long time? If do, I expect a unittest added to make sure NodeReady is the last one in status list.

**Comment by pmorie** [comment]
Thanks for the PR, @aaronlevy

**Comment by aaronlevy** [comment]
I added a sanity check to the existing nodestatus unit tests to ensure nodeReady was the last condition, but I can break that out into another unit test if necessary.

Also, any chance this could be cherry-picked into the v1.1 releases? Should I open a separate PR for that?

**Comment by dchen1107** [comment]
Ahh, I missed that unittest change. Ok that should be ok. 

You can run ./hack/cherry_pick_pull.sh to cherry pick this pr to 1.1 branch. But please note that we shouldn't run into this issue based on today's Kubernetes cluster upgrade policy. In this case, we are not going to cut a release for this pr only. cc/ @quinton-hoole @mikedanese @brendandburns

**Comment by wojtek-t** [comment]
@aaronlevy - can you please rebase this PR?

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -2873,6 +2827,55 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 		node.Status.Conditions = append(node.Status.Conditions, *nodeOODCondition)
 	}
 
+	// NOTE(aaronlevy): NodeReady condition needs to be the last in the list of node conditions.
+	// This is due to an issue with version skewed kubelet and master components.
+	// ref: https://github.com/kubernetes/kubernetes/issues/16961
```
Comment (mikedanese) [inline pkg/kubelet/kubelet.go]: cc @ihmccreery

**Pull request merged**